lib/si-split-graph/src/subgraph.rs

```diff
@@ -103,42 +103,37 @@ where
         subgraph
     }
 
-    pub(crate) fn cleanup(&mut self) -> Vec<SplitGraphNodeId> {
-        loop {
-            let orphaned_node_indexes: Vec<SubGraphNodeIndex> = self
-                .graph
-                .externals(Incoming)
-                .filter(|idx| *idx != self.root_index)
-                .collect();
-
-            if orphaned_node_indexes.is_empty() {
-                break;
+    pub(crate) fn remove_externals(&mut self) -> Vec<SplitGraphNodeId> {
+        let mut removed_ids = vec![];
+        let mut indexes_to_remove = vec![];
+        for external in self
+            .graph
+            .externals(Incoming)
+            .filter(|idx| *idx != self.root_index)
+        {
+            if let Some(node_id) = self.graph.node_weight(external).map(|node| node.id()) {
+                removed_ids.push(node_id);
             }
+            indexes_to_remove.push(external);
+        }
 
-            for node_index in orphaned_node_indexes {
-                self.graph.remove_node(node_index);
-            }
+        for index in indexes_to_remove {
+            self.graph.remove_node(index);
         }
 
-        let mut removed_ids = vec![];
+        removed_ids
+    }
```

Comment:
Won't this need to loop until it stops finding externals (that aren't the root node)? It looks like it would only remove the first "layer" of externals, and not anything that becomes an external when that first set is removed.

Reply:
Yes, that happens at the next level up, in the cleanup function for the splitgraph itself. It loops over all the subgraphs, removing externals, then removing the external source edges, and repeating.

Reply:
Yeah, I saw that `SplitGraph::cleanup` looped. Even though this is `pub(crate)`, if it's not going to handle its own looping, it should probably at least have a doc comment basically saying "Don't use this, unless you know exactly what you're doing. Go use `SplitGraph::cleanup` instead.".

Reply:
@jhelwig i've added a doc comment to that effect